lib/vmpooler/pool_manager.rb

```diff
@@ -619,8 +619,21 @@ def load_used_providers
     # @return [Array] - a list of used providers from the config file, defaults to the default providers
     # ie. ["dummy"]
     def used_providers
-      pools = config[:pools] || []
-      @used_providers ||= (pools.map { |pool| pool[:provider] || pool['provider'] }.compact + default_providers).uniq
+      # create an array of provider classes based on the config
+      if config[:providers]
+        config_provider_names = config[:providers].keys
+        config_providers = config_provider_names.map do |config_provider_name|
+          if config[:providers][config_provider_name] && config[:providers][config_provider_name]['provider_class']
+            config[:providers][config_provider_name]['provider_class']
+          else
+            config_provider_name
+          end
+        end.compact.uniq
+      else
+        config_providers = []
+      end
```

Comment:
Nice use of map, which made me think we can one line it with the .map() for hashes which return an array
```suggestion
 config_providers = config[:providers].map { |k,v| v['provider_class'] || k.to_s }.compact.uniq
```

Reply:
causes `undefined method `[]' for nil:NilClass` to be thrown on basically every test

Reply:
I think I see why...

Reply:
nm. I think there are too many places that the key might not exist to work with that map you posted

Reply:
config_providers = config[:providers].map { |k,v| v['provider_class'] || k.to_s }.compact.uniq if config[:providers]

Reply:
just missing the case where its not set (which is probably a lot of tests)

Reply:
mabe keep the if config[:providers] else blocks

Reply:
sadly, still blows up

Reply:
I am still going to look at this a little more though, I like what you've started

Reply:
It seems there are too many places things can be blank to shrink this down more. I think I am just going to leave it for now.